alf/algorithms/sac_algorithm.py

```diff
@@ -190,53 +194,80 @@ def predict(self, time_step: ActionTimeStep, state=None):
             network_state=state)
         return PolicyStep(action=action, state=state, info=())
 
-    def _actor_train_step(self, exp: Experience, state: SacActorState, action,
-                          log_pi):
-        critic_input = (exp.observation, action)
-
-        with tf.GradientTape(watch_accessed_variables=False) as tape:
-            tape.watch(action)
+    def _actor_train_step(self, exp: Experience, state: SacActorState,
+                          action_distribution, action, log_pi):
+
+        if self._is_continuous:
+            critic_input = (exp.observation, action)
+
+            with tf.GradientTape(watch_accessed_variables=False) as tape:
+                tape.watch(action)
+                critic1, critic1_state = self._critic_network1(
+                    critic_input,
+                    step_type=exp.step_type,
+                    network_state=state.critic1)
+
+                critic2, critic2_state = self._critic_network2(
+                    critic_input,
+                    step_type=exp.step_type,
+                    network_state=state.critic2)
+
+                target_q_value = tf.minimum(critic1, critic2)
+
+            dqda = tape.gradient(target_q_value, action)
+
+            def actor_loss_fn(dqda, action):
+                if self._dqda_clipping:
+                    dqda = tf.clip_by_value(dqda, -self._dqda_clipping,
+                                            self._dqda_clipping)
+                loss = 0.5 * losses.element_wise_squared_loss(
+                    tf.stop_gradient(dqda + action), action)
+                loss = tf.reduce_sum(
+                    loss, axis=list(range(1, len(loss.shape))))
+                return loss
+
+            actor_loss = tf.nest.map_structure(actor_loss_fn, dqda, action)
+            alpha = tf.stop_gradient(tf.exp(self._log_alpha))
+            actor_loss += alpha * log_pi
+        else:
             critic1, critic1_state = self._critic_network1(
-                critic_input,
+                exp.observation,
                 step_type=exp.step_type,
                 network_state=state.critic1)
 
             critic2, critic2_state = self._critic_network2(
-                critic_input,
+                exp.observation,
                 step_type=exp.step_type,
                 network_state=state.critic2)
 
-            target_q_value = tf.minimum(critic1, critic2)
-
-        dqda = tape.gradient(target_q_value, action)
-
-        def actor_loss_fn(dqda, action):
-            if self._dqda_clipping:
-                dqda = tf.clip_by_value(dqda, -self._dqda_clipping,
-                                        self._dqda_clipping)
-            loss = 0.5 * losses.element_wise_squared_loss(
-                tf.stop_gradient(dqda + action), action)
-            loss = tf.reduce_sum(loss, axis=list(range(1, len(loss.shape))))
-            return loss
+            action_probs = action_distribution._probs
```

Comment:
action_distribution._probs is None when I run the code using tensorflow 2.0.0-beta1. Somehow it is not converted from logits.
Result from print(action_distribution.__dict__):
{'_parameters': {'name': 'Categorical', 'allow_nan_stats': True, 'validate_args': False, 'dtype': tf.int64, 'probs': **None**, 'logits': <tf.Tensor: id=509, shape=(8, 2), dtype=float32, numpy=
array([[ 1.6195892e-03, -4.9218588e-04], ....

Reply:
distributions it's in package `tfp` not in `tensorflow` , i use `tfp-nightly==0.8.0.dev20190609` in my workstation 

and avoid dependency inconsistency,  i suggest use docker as our dev env
[alf image gpu](https://docker.hobot.cc/harbor/projects/14/repositories/rl%2Ftfagent)
[alf image cpu](https://docker.hobot.cc/harbor/projects/14/repositories/rl%2Ftfagent-cpu)

Reply:
Tested with this version of tfp and works good.